### PR TITLE
feat(dart/transform): Inline `templateUrl` values

### PR DESCRIPTION
**Pull request by kegluneq**
Modify DirectiveProcessor to inline `templateUrl` values to avoid making
additional browser requests.

### CODE REVIEW

**Comment by kegluneq** [comment]
This is not quite fully baked yet, but thought I'd send it out for some early thoughts. I won't be looking at this over the weekend, so no hurry at all.

**Comment by kegluneq** [comment]
Alright, added some additional comments and tests -- I think this is ready for review.

Thanks!

File: modules/angular2/src/transform/common/async_string_writer.dart [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+
+/// [PrintWriter] implementation that allows asynchronous printing via
+/// [asyncPrint] and [asyncToString]. See those methods for details.
+class AsyncStringWriter extends PrintWriter {
+  /// All [Future]s we are currently waiting on.
+  final List<Future<String>> _awaiting = <Future<String>>[];
+  final List<StringBuffer> _bufs;
+  StringBuffer _curr;
+
+  AsyncStringWriter._(StringBuffer curr)
```
Comment (yjbanov) [inline modules/angular2/src/transform/common/async_string_writer.dart]: Do you need this private constructor? It's only used once.

Comment (yjbanov) [inline modules/angular2/src/transform/common/async_string_writer.dart]: Oh, I see this is so you can make `_bufs` final. nvm then.

File: modules/angular2/src/transform/common/async_string_writer.dart [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+        .whenComplete(() => _awaiting.remove(futureText));
+  }
+
+  /// Waits for any values added via [asyncPrint] and returns the fully
+  /// built string.
+  Future<String> asyncToString() async {
+    // Save length in case it is updated while we wait.
+    var bufLen = _bufs.length;
+    if (bufLen == 1) return '$_curr';
+    await Future.wait(_awaiting);
```
Comment (yjbanov) [inline modules/angular2/src/transform/common/async_string_writer.dart]: For robustness this should probably be:

```dart
while(_awaiting.length > 0) {
  await Future.wait(_awaiting);
}
```

Although, actually, I'm not sure what `Future.wait` does if the contents of the list change in-flight.

Comment (kegluneq) [inline modules/angular2/src/transform/common/async_string_writer.dart]: `Future.wait` is smart enough to handle that, but it turns out my code wasn't. It's fixed now.

Thanks!

**Comment by kegluneq** [comment]
Squashed + merged in 97d24563f492b5cea29e8da3e8e41c69d3d10c95

**Pull request closed without merging**